lib/luaxtable.c

```diff
@@ -32,48 +32,125 @@
 #include <lauxlib.h>
 #include <lunatik.h>
 
+#include "luadata.h"
+#include "luarcu.h"
+#include "luaxtable.h"
+
 typedef enum luaxtable_type_e {
 	LUAXTABLE_TMATCH,
 	LUAXTABLE_TTARGET,
 } luaxtable_type_t;
 
 typedef struct luaxtable_s {
 	lunatik_object_t *runtime;
+	lunatik_object_t *skb;
 	union {
 		struct xt_match match;
 		struct xt_target target;
 	};
+	int fallback;
 	luaxtable_type_t type;
 } luaxtable_t;
 
-static bool luaxtable_match(const struct sk_buff *skb, struct xt_action_param *par)
-{
-	return true;
-}
+static struct {
+	lunatik_object_t *target;
+	lunatik_object_t *match;
+} luaxtable_hooks = {NULL, NULL};
 
-static int luaxtable_match_check(const struct xt_mtchk_param *par)
+static int luaxtable_invoke(lua_State *L, luaxtable_t *xtable, struct sk_buff *skb, const char *op, int nret)
 {
-	return 0;
-}
+	int base = lua_gettop(L);
+	int ret = -ENXIO;
+	int nargs = 0;
+	lunatik_object_t *data = NULL;
+
+	if (lunatik_getregistry(L, xtable) != LUA_TTABLE) {
+		pr_err("could not find ops table\n");
+		goto err;
+	}
 
-static void luaxtable_match_destroy(const struct xt_mtdtor_param *par)
-{
+	if (lua_getfield(L, -1, op) != LUA_TFUNCTION) {
+		pr_err("%s isn't defined\n", op);
+		goto err;
+	}
+
+	if (skb != NULL ){
+		if (lunatik_getregistry(L, xtable->skb) != LUA_TUSERDATA) {
+			pr_err("could not find skb\n");
+			goto err;
+		}
+
+		data = (lunatik_object_t *)lunatik_toobject(L, -1);
+		luadata_reset(data, skb, sizeof(struct sk_buff));
+		nargs = 1;
+	}
+
+	if (lua_pcall(L, nargs, nret, 0) != LUA_OK) {
+		pr_err("%s error: %s\n", op, lua_tostring(L, -1));
+		ret = -ECANCELED;
+		goto err;
+	}
+	ret = luaL_optinteger(L, -1, 0);
+
+err:
+	if (data)
+		luadata_clear(data);
+
+	if (ret < 0)
+		pr_err("error in running %s lua handler : %d\n", op, ret);
```

Comment:
you don't need this extra print here, right? you can just add the `op` info to every error condition.. you can even have a macro for this..